engine/src/main/java/nl/inl/blacklab/indexers/config/saxon/XPathFinder.java

```diff
@@ -41,38 +47,49 @@ public class XPathFinder {
     /** Variables to make available from XPath */
     private final Map<String, String> vars = new HashMap<>();
 
-    public XPathFinder(Map<String, String> namespaces, Map<String, String> vars) {
-        this.vars.putAll(vars);
-
-        // Set up namespace aware xpath that will compile xpath expressions
-        xPath = SaxonHelper.newXPathFactory();
-        xPath.setCaching(true); // cache xpaths
+    private final LoadingCache<Map<String, String>, XPathCompiler> compilerCache = Caffeine.newBuilder()
+        .maximumSize(50) // should be large enough for most use cases?
+        .expireAfterAccess(Duration.ofMinutes(1))
+        .build(namespaces -> {
+            var fac = SaxonHelper.newXPathFactory();
+            fac.setCaching(true);
+            
+            for (String var: vars.keySet()) {
+                fac.declareVariable(new QName(var));
+            }
+            // xml namespace is implicit
+            fac.declareNamespace(NAMESPACE_XML_PREFIX, NAMESPACE_XML_URI);
+            boolean hasNamespaces = false;
+            if (namespaces != null && !namespaces.isEmpty()) {
+                for (Map.Entry<String, String> e: namespaces.entrySet()) {
+                    if (e.getKey().equals(NAMESPACE_XML_PREFIX)) {
+                        if (!e.getValue().equals(NAMESPACE_XML_URI))
+                            logger.warn("Tried to redefine implicit 'xml' namespace prefix to '" + e.getValue()
+                                    + "'); ignoring");
+                        continue;
+                    }
+                    hasNamespaces = true;
+                    fac.declareNamespace(e.getKey(), e.getValue());
+                }
+            }
+            if (!hasNamespaces) {
+                // No namespaces declared in the indexer config.
+                // Set Saxon to ignore namespace on elements without a prefix.
+                // This makes sure that we can index documents with or without namespaces, which
+                // unfortunately sometimes happens in large corpora.
+                fac.setUnprefixedElementMatchingPolicy(UnprefixedElementMatchingPolicy.ANY_NAMESPACE);
+            }
+            return fac;
+        });
 
-        for (String var: vars.keySet()) {
-            xPath.declareVariable(new QName(var));
-        }
+    private static final ThreadLocal<Map<String, XPathSelector>> compiledXPaths = ThreadLocal.withInitial(java.util.HashMap::new);
 
-        // xml namespace is implicit
-        xPath.declareNamespace(NAMESPACE_XML_PREFIX, NAMESPACE_XML_URI);
+    public XPathFinder(Map<String, String> namespaces, Map<String, String> vars) {
```

Comment:
Can we safely cache compiled XPaths? XpathSelector depends on vars, which change for each document indexed.